nntrainer/cl_context.cpp

```
@@ -219,45 +251,46 @@ bool ClContext::clCreateKernel(std::string &kernel_string,
-      size_t binary_size = fs.tellg();
-      fs.seekg(0, std::ios::beg);
-
-      std::vector<unsigned char> chunk(binary_size);
-      fs.read((char *)chunk.data(), binary_size);
-
-      result = program.CreateCLProgramWithBinary(
-        opencl::ContextManager::Global().GetContext(),
-        opencl::ContextManager::Global().GetDeviceId(), binary_size,
-        chunk.data(),
-        opencl::Program::DEFAULT_KERNEL_PATH + "/" + kernel_name +
-          "_kernel.bin",
-        "");
-    } else {
-      result = program.CreateCLProgram(
-        opencl::ContextManager::Global().GetContext(),
-        opencl::ContextManager::Global().GetDeviceId(), kernel_string, "");
-    }
-
-    if (!result) {
-      break;
+  opencl::Program program;
+
+  // reading binary
+  std::string binary_file_path =
+    opencl::Program::DEFAULT_KERNEL_PATH + "/" +
+    std::to_string(program.GetKernelHash(kernel_string, "")) + ".cl.bin";
+  auto binary_data = readBinaryFile(binary_file_path);
+
+  if (!binary_data.empty()) {
+    ml_logi("Using cached version of kernel: %s at path %s",
+            kernel_name.c_str(), binary_file_path.c_str());
+    result = program.CreateCLProgramWithBinary(
+      opencl::ContextManager::Global().GetContext(),
+      opencl::ContextManager::Global().GetDeviceId(), binary_data,
+      binary_file_path, "");
+  } else {
+    ml_logi("Binary for kernel %s not found, compiling from source...",
+            kernel_name.c_str());
+    result = program.CreateCLProgram(
+      opencl::ContextManager::Global().GetContext(),
+      opencl::ContextManager::Global().GetDeviceId(), kernel_string, "");
+
+    if (result) {
+      auto binary = program.GetProgramBinary(
+        opencl::ContextManager::Global().GetDeviceId());
+
+      if (binary.empty()) {
+        ml_loge("Failed retrieving binary for kernel %s", kernel_name.c_str());
+        result = false;
+      } else {
+        result &= writeBinaryFile(binary_file_path, binary);
+      }
```

I think we need to properly handle the error here when the result is false. When there's a kernel with syntax errors, it consistently tries to compile from the source. Let's throw an error when `CreateCLProgram` fails.

---

Yes, something like below

```C++
if (!result) {
  throw std::runtime_error(
    "Error: Failed to create OpenCL program " + kernel_name);
}
```

---

+ It would be nice to show what the problem is as follows.

https://github.com/nnstreamer/nntrainer/blob/bce948a22a57eab14524f6f856006585bcc3c9f0/nntrainer/opencl/opencl_program.cpp#L48-L51

---

Let's not use `ml_loge` since it only logs the error message.

---

Sounds good :)